lib/compilers/typescript-native.ts

```diff
@@ -41,107 +41,83 @@ export class TypeScriptNativeCompiler extends BaseCompiler {
     }
 
-    override optionsForFilter(filters: ParseFiltersAndOutputOptions, outputFilename: string) {
-        return [this.filename(outputFilename)];
-    }
-
-    override async handleInterpreting(key, executeParameters) {
-        executeParameters.args = [
-            '--emit=jit',
-            this.tscSharedLib ? '--shared-libs=' + this.tscSharedLib : '-nogc',
-            ...executeParameters.args,
-        ];
+    override optionsForFilter(
+        filters: ParseFiltersAndOutputOptions,
+        outputFilename: string,
+        userOptions?: string[],
+    ): string[] {
+        return [];
+    }    
 
-        return await super.handleInterpreting(key, executeParameters);
-    }
+    override optionsForBackend(backendOptions: Record<string, any>, outputFilename: string): string[] {
+        let addOpts: string[] = [];
 
-    override async runCompiler(
-        compiler: string,
-        options: string[],
-        inputFilename: string,
-        execOptions: ExecutionOptions,
-    ): Promise<CompilationResult> {
-        // These options make Clang produce an IR
-        const newOptions = ['--emit=mlir-llvm', inputFilename];
+        addOpts.push(this.tscAsmOutput ? '--emit=asm' : '--emit=mlir');
 
         if (!this.tscSharedLib) {
-            newOptions.push('-nogc');
+            addOpts.push('-nogc');
         }
 
-        const output = await this.runCompilerRawOutput(
-            this.tscJit,
-            newOptions,
-            this.filename(inputFilename),
-            execOptions,
-        );
-        if (output.code !== 0) {
-            return {
-                code: output.code,
-                timedOut: false,
-                stdout: [],
-                stderr: [
-                    {
-                        text: 'Failed to run compiler to get MLIR code',
-                    },
-                ],
-            };
+        if (this.tscNewOutput) {
+            addOpts.push('-o=' + this.filename(outputFilename));
         }
 
-        return {code: 0, timedOut: false, stdout: [], stderr: []};
+        return addOpts;
+    }    
+
+    override getIrOutputFilename(inputFilename: string, filters: ParseFiltersAndOutputOptions): string {
+        const outputFilename = this.getOutputFilename(path.dirname(inputFilename), this.outputFilebase);
+        // As per #4054, if we are asked for binary mode, the output will be in the .s file, no .ll will be emited
+        if (!filters.binary) {
+            return outputFilename.replace('.s', '.ll');
+        }
+        return outputFilename;
     }
 
-    override async generateIR(
-        inputFilename: string,
-        options: string[],
-        irOptions: LLVMIrBackendOptions,
-        filters: ParseFiltersAndOutputOptions,
-    ) {
-        // These options make Clang produce an IR
-        let newOptions = ['--emit=llvm', inputFilename];
+    override async generateIR(inputFilename: string, options: string[], filters: ParseFiltersAndOutputOptions) {
```

Comment:
This new function signature seems like a mistake, and I'm surprised it compiles. The third argument isn't of type `ParseFiltersAndOutputOptions` in the base compiler.

Reply:
updated, can u review it again, pls